### PR TITLE
Robot Software: Diagnostics, Ramping, Calibration

### LINKED ISSUE #1201: Visual Indication for when e-stop is on (RobotDiagnostics)
### Description of the task

After several occurrences of thinking robots were broken when the e-stop was just on without us realizing, Jade made the excellent suggestion that we give a Visual cue in our GUI tools that indicate the e-stop is on. 

Options are:
* Make everything red (but still readable)
* Something else? Open to ideas

### Acceptance criteria

- [ ] Add a visual indication to the Robot Diagnostics tool and AI Visualizer that indicate when the e-stop is on

### LINKED ISSUE #2112: Update robot constants
### Description of the task
Update 2021 robot constants; need to talk to mech/elec and/or do field testing

### Acceptance criteria

- [ ] update robot constants, see TODOs

### LINKED ISSUE #2563: Thunderscope: Robot Diagnostics Integration

### LINKED ISSUE #2594: Flashing ESP32 from the Jetson Nano
We need to flash the esp32 from the jetson over the uart connection we use to communicate with power loop.

### LINKED ISSUE #2644: Integrate RobotStatus msgs with Thunderscope (per-wheel velocities, target velocity, temps, etc..)

### PR DESCRIPTION
**Pull request by akhilveeraghanta**
### Resolved Issues
resolves #2594
resolves #1201 
resolves #2112
resolves #2644
resolves #2563

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

File: src/extlibs/hrvo/hrvo_agent.cpp
Comment (itsarune) [inline src/extlibs/hrvo/hrvo_agent.cpp]: I don't think you meant to commit this change

File: src/extlibs/hrvo/simulator.cpp
Comment (itsarune) [inline src/extlibs/hrvo/simulator.cpp]: You created "velocity" but you didn't end up using it

File: src/software/jetson_nano/primitive_executor.cpp
Comment (itsarune) [inline src/software/jetson_nano/primitive_executor.cpp]: Do we have a conclusion on what we're going to do for angular correction?

File: src/software/jetson_nano/display/robot_ui.py
Comment (nimazareian) [inline src/software/jetson_nano/display/robot_ui.py:82]: Curious why we need the sleep here?

File: src/software/jetson_nano/linux_configs/extlinux.conf
Comment (nimazareian) [inline src/software/jetson_nano/linux_configs/extlinux.conf:17]: Curious what this file is for?

File: src/software/jetson_nano/primitive_executor.cpp
Comment (nimazareian) [inline src/software/jetson_nano/primitive_executor.cpp:77]: This change doesn't seem necessary

File: src/software/jetson_nano/primitive_executor.h
Comment (nimazareian) [inline src/software/jetson_nano/primitive_executor.h:76]: This method doesn't exist anymore, so can be deleted.

File: src/software/jetson_nano/primitive_executor.h
Comment (nimazareian) [inline src/software/jetson_nano/primitive_executor.h:92]: I think this is also unused (it should also be contained in the Primitive)

File: src/software/jetson_nano/services/motor.cpp
Comment (nimazareian) [inline src/software/jetson_nano/services/motor.cpp:67]: This should be moved to robot constants, or be tracked with a ticket

File: src/software/jetson_nano/services/motor.cpp
Comment (nimazareian) [inline src/software/jetson_nano/services/motor.cpp:777]: Curious if this still applies? Does the robot have to be lifted off of the ground, or can we do it on the ground?

File: src/software/jetson_nano/services/motor.cpp
Comment (nimazareian) [inline src/software/jetson_nano/services/motor.cpp:290]: Is this log necessary, or was it added for debugging? Seems a bit spammy

File: src/software/jetson_nano/services/motor.cpp
Comment (nimazareian) [inline src/software/jetson_nano/services/motor.cpp:405]: I think our final testings with ramping had the rampings separated (i.e. ramp wheel velocities calculated from linear velocities, ramp velocities calculated from angular velocity, add the resultant vectors). Though just thinking about it, it makes more sense to do the ramping once, since adding the two resultant vectors could potentially have the delta wheel velocities greater than what we want to limit it to.

We should figure out some max delta velocity (i.e. max wheel acceleration) threshold (not sure how we can pick a realistic/safe number) and max wheel speed, then use those values for ramping. Instead of using the ones currently in the robot constants as those represent the euclidean velocity limits rather than per wheel.

File: src/software/jetson_nano/services/motor.cpp
Comment (nimazareian) [inline src/software/jetson_nano/services/motor.cpp:367]: not exactly sure if this was added for debugging or if it's meant to exist here. Seems like we're using it as a way to ramp the dribbler rpm rather than just set it to the requested rpm? We could probably implement it with out a static int.

File: src/software/jetson_nano/services/motor.cpp
Comment (nimazareian) [inline src/software/jetson_nano/services/motor.cpp:385]: This was explained to me previously, but Im still confused on why we need to do this conversion. We should either remove it (Don't think it will actually effect this method) or add a comment explaining why we need it.

File: src/software/jetson_nano/services/motor.cpp
Comment (nimazareian) [inline src/software/jetson_nano/services/motor.cpp:456]: Should be updated to the newer version which includes limiting the wheel velocities as well:
Something like:
```cpp
    // find absolute max wheel velocity
    auto max_ramp_wheel_velocity = ramp_wheel_velocity.cwiseAbs().maxCoeff();

    // compare against max wheel velocity
    if (max_ramp_wheel_velocity > max_allowable_wheel_velocity)
    {
        // if larger, scale down to max
        ramp_wheel_velocity = (ramp_wheel_velocity / max_ramp_wheel_velocity) *
                              max_allowable_wheel_velocity;
    }
```

File: src/software/jetson_nano/services/motor.cpp
Comment (nimazareian) [inline src/software/jetson_nano/services/motor.cpp:772]: I'm unfamiliar with how the motor/motor controllers work, but just curious how you went about determining these hexadecimal values (e.g. `0x00000FFF`). We should consider using variables to make it a bit clearer what each `writeToControllerOrDieTrying` is doing.

File: src/software/jetson_nano/services/motor.cpp
Comment (nimazareian) [inline src/software/jetson_nano/services/motor.cpp:758]: @itsarune @akhilveeraghanta Can we close #2451?

File: src/software/jetson_nano/services/motor.h
Comment (nimazareian) [inline src/software/jetson_nano/services/motor.h:210]: The comment and variable name here seem a bit counter intuitive

File: src/software/jetson_nano/services/network.cpp
Comment (nimazareian) [inline src/software/jetson_nano/services/network.cpp:43]: nit: Should this be 
```cpp
total_primitives_lost += input.sequence_number() - last_primitive_sequence_number - 1;
```
instead?

File: src/software/jetson_nano/services/network.cpp
Comment (nimazareian) [inline src/software/jetson_nano/services/network.cpp:52]: An idea for the future: It could also be helpful to have some logic to check that the primitives are being received at a consistent rate. As in there is not a big gap between two subsequent primitives. Not sure if it is possible, but mainly worried of some sort of networking lag that results in us getting the primitive, but receiving it after a long time that it might be invalid. Same with the World messages.
What do you think?
(Edit: seems like we already have `last_primitive_time` and `last_world_time`, but we don't use them)

File: src/software/jetson_nano/thunderloop.cpp
Comment (nimazareian) [inline src/software/jetson_nano/thunderloop.cpp:121]: Comment can be removed.

File: src/software/jetson_nano/thunderloop.cpp
Comment (nimazareian) [inline src/software/jetson_nano/thunderloop.cpp:186]: @jonathanlew Not really sure what this TODO refers to? We do pass the World to primitive executor using the `updateWorld` call above.

File: src/software/jetson_nano/thunderloop.cpp
Comment (nimazareian) [inline src/software/jetson_nano/thunderloop.cpp:207]: The debug logs can be removed

File: src/software/jetson_nano/thunderloop.cpp
Comment (nimazareian) [inline src/software/jetson_nano/thunderloop.cpp:244]: This debug log + `loop_duration_seconds` can also be removed I think

File: src/software/physics/BUILD
Comment (nimazareian) [inline src/software/physics/BUILD:13]: Can remove this dependency and remove the #include in the implementation.

File: src/software/thunderscope/robot_communication.py
Comment (nimazareian) [inline src/software/thunderscope/robot_communication.py:45]: Might need a `HRVOVisualization` buffer as well. Same with a `HRVOVisualizationProtoListener` lower down.

File: src/software/thunderscope/robot_communication.py
Comment (nimazareian) [inline src/software/thunderscope/robot_communication.py:89]: I believe from our previous testing we ended up having to initialize `self.fullsystem_connected_to_robots` to False or else the variable wouldn't be initialized before this read

File: src/software/thunderscope/robot_diagnostics/chicker.py
Comment (nimazareian) [inline src/software/thunderscope/robot_diagnostics/chicker.py:119]: nit: add issue number
Also, not sure about this todo. If we don't want auto (assuming auto kick), we want to kick immediately?

File: src/software/thunderscope/robot_communication.py
Comment (nimazareian) [inline src/software/thunderscope/robot_communication.py:192]: Thanks for noting this. I've created #2711 for now. I'm not sure if I fully understood the task and explained it properly. Feel free to update it.
Also, curious when `self.disconnect_fullsystem_from_robots()` should be used and if its necessary?

**Review by nimazareian — CHANGES_REQUESTED** [review]
Thanks for taking the time to organize the branch and create this PR. I know that the branch was a mess after robocup.
I have some questions and changes I've requested. I understand that you're probably already very busy, so we can have some people on the team make the changes on your branch or create a PR to your branch with the changes.

File: src/shared/2021_robot_constants.cpp
Comment (nimazareian) [inline src/shared/2021_robot_constants.cpp:11]: Not sure if these values are all up to date now, but we should probably remove the todo for the ones that are up to date, and if there are any that are still estimates which need to be updated, we should probably not resolve #2112.

**Review by nimazareian — CHANGES_REQUESTED** [review]

File: src/software/jetson_nano/display/screens/home_screen.py
Comment (itsarune) [inline src/software/jetson_nano/display/screens/home_screen.py:128]: docs for what x or y are?

File: src/software/jetson_nano/services/motor.cpp
Comment (itsarune) [inline src/software/jetson_nano/services/motor.cpp:404]: attach a TODO issue number

File: src/software/jetson_nano/services/motor.cpp
Comment (itsarune) [reply src/software/jetson_nano/services/motor.cpp:758]: at some point, we need to look into calibration with absolute encoders

File: src/software/jetson_nano/services/motor.h
Comment (itsarune) [inline src/software/jetson_nano/services/motor.h:63]: docs

File: src/software/jetson_nano/services/motor.h
Comment (itsarune) [inline src/software/jetson_nano/services/motor.h:107]: missing @param newline

File: src/software/jetson_nano/services/motor.h
Comment (itsarune) [inline src/software/jetson_nano/services/motor.h:138]: param missing docs

File: src/software/jetson_nano/services/network.cpp
Comment (itsarune) [reply src/software/jetson_nano/services/network.cpp:52]: this makes sense to me. where would you put this? thunderloop?

File: src/software/jetson_nano/services/power.h
Comment (itsarune) [inline src/software/jetson_nano/services/power.h:53]: does this not work as a static consexpr

File: src/software/thunderscope/robot_communication.py
Comment (itsarune) [reply src/software/thunderscope/robot_communication.py:45]: doesn't need to be included in this PR, I'll add it later on as part of the robot visualization from robot PR

File: src/software/physics/euclidean_to_wheel_test.cpp
Comment (itsarune) [inline src/software/physics/euclidean_to_wheel_test.cpp:27]: why're we deleting all these tests?

**Review by itsarune — COMMENTED** [review]
his final word, we'll talk about handing over the responsibility for this PR

File: src/software/jetson_nano/services/power.h
Comment (LiCody) [reply src/software/jetson_nano/services/power.h:53]: i believe you cant do constexpr with std::string

**Review by LiCody — COMMENTED** [review]
Looks like some updates to powerloop are missing see: https://github.com/LiCody/Software/tree/cody/change_baud/src/software/power . Above changes also need some clean up so I will add them once this PR is merged

File: src/software/physics/euclidean_to_wheel_test.cpp
Comment (nimazareian) [reply src/software/physics/euclidean_to_wheel_test.cpp:27]: I've rewritten pretty much all of these tests and will include it in a future PR once Im more comfortable with the conversion. I mainly rewrote the tests as they were too tightly coupled with the wheel angle.

File: src/software/jetson_nano/services/network.cpp
Comment (nimazareian) [reply src/software/jetson_nano/services/network.cpp:52]: I would probably keep it here given that its more directly related to the networking. But the advantage of it being in Thunderloop could be that if we completely lose connection and don't receive primitives for a while, we could have some logic that (e.g.) resets the primitive to have stop command and logs that we havent received a primitive in X time. But if the logic is here, it only gets called when a new primitive is received, so we wouldnt know/log if we don't receive a primitive for a while.

**Comment by LiCody** [comment]
closed in favour of https://github.com/UBC-Thunderbots/Software/pull/2721

**Pull request closed without merging**
